google/cloud/pubsub_v1/subscriber/_protocol/dispatcher.py

```diff
@@ -168,17 +168,30 @@ def ack(self, items: Sequence[requests.AckRequest]) -> None:
 
         # We must potentially split the request into multiple smaller requests
         # to avoid the server-side max request size limit.
-        ack_ids = (item.ack_id for item in items)
+        items_gen = (i for i in items)
+        ack_ids_gen = (item.ack_id for item in items)
         total_chunks = int(math.ceil(len(items) / _ACK_IDS_BATCH_SIZE))
 
         for _ in range(total_chunks):
-            request = gapic_types.StreamingPullRequest(
-                ack_ids=itertools.islice(ack_ids, _ACK_IDS_BATCH_SIZE)
-            )
-            self._manager.send(request)
-
-        # Remove the message from lease management.
-        self.drop(items)
+            future_reqs_dict = {req.ack_id: req for req in itertools.islice(items_gen, _ACK_IDS_BATCH_SIZE) if req.future}
+            requests_completed, requests_to_retry =  self._manager.send_unary_ack(
+                ack_ids=list(itertools.islice(ack_ids_gen, _ACK_IDS_BATCH_SIZE)),
+                future_reqs_dict=future_reqs_dict)
+
+            # Remove the completed messages from lease management.
+            self.drop(requests_completed)
+
+            # retry acks
+            # todo: figure out if retry delays/limits are needed
+            while requests_to_retry:
```

Comment:
How long can this loop spin around?

If there is a transient error, a failed unary ACK request could be sent many times in a quick succession here, which is undesirable.

In addition, the `ack()` method could block for a long time while eagerly retrying, effectively blocking the `dispatch_callback()` method and the entire dispatcher. This means, for instance, that any new lease requests added to the dispatch queue will not be processed in time and leased messages could unnecessarily expire.

If we really want to repeatedly make requests here, it should be delegated to a separate thread.

Reply:
Also applies to the `drop()` method, which is very similar.

(we could also consider extracting some of the common logic out, if it does not hinder readability)

Reply:
@pradn This seems to have slipped through, are we sure this will behave gracefully in all cases? Won't it block the thread for too long?

Reply:
I plan to fix this. I'd like to get some initial testing done before. (Should have mentioned this to you :) )

Reply:
OK good then. :+1: For a moment I got worried that we forgot about this. :)

Reply:
Fixed: https://github.com/googleapis/python-pubsub/pull/550/commits/d7305e292defc509762f163c7ed2ea4ac87a73d0